sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java

```diff
@@ -0,0 +1,143 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugin.typescript;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.sonar.api.Startable;
+import org.sonar.api.batch.InstantiationStrategy;
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.plugin.typescript.executable.ExecutableBundle;
+import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+@InstantiationStrategy("PER_PROCESS")
+@SonarLintSide
+public class ContextualServer implements Startable {
+
+  // SonarLint should pass in this property an absolute path to the directory containing TypeScript dependency
+  private static final String TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY = "sonar.typescript.internal.typescriptLocation";
+
+  private static final Logger LOG = Loggers.get(ContextualServer.class);
+
+  private Configuration configuration;
+  private ExecutableBundleFactory bundleFactory;
+  private static AtomicReference<Process> serverProcess = new AtomicReference<>();
+
+  public ContextualServer(ExecutableBundleFactory bundleFactory) {
+    this(null, bundleFactory);
+  }
+
+  public ContextualServer(Configuration configuration, ExecutableBundleFactory bundleFactory) {
+    this.configuration = configuration;
+    this.bundleFactory = bundleFactory;
+  }
+
+  @Override
+  public void start() {
+    LOG.info("Starting SonarTS Server");
+
+    if (serverProcess.get() != null && serverProcess.get().isAlive()) {
+      LOG.warn("Skipping SonarTS Server start, already running");
+      return;
+    }
+
+    if (configuration == null) {
+      LOG.warn("Skipping SonarTS Server start due to missing configuration");
+      return;
+    }
+    
+    startSonarTSServer();
+  }
+
+  private void startSonarTSServer() {
+    ExecutableBundle bundle = bundleFactory.createAndDeploy(getServerDir(), configuration);
+
+    ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand().commandLineTokens());
+    setNodePath(processBuilder);
+
+    try {
+      processBuilder.inheritIO();
+      serverProcess.set(processBuilder.start());
+      //TODO find a better way to wait until the server is up
+      // e.g. loop with accessing the port
+      synchronized (this) {
+        wait(1000);
+      }
+      LOG.info("SonarTS Server is started");
+
+    } catch (IOException e) {
+      LOG.error("Failed to start SonarTS Server", e);
+
+    } catch (InterruptedException e) {
+      LOG.error("Failed to wait until the SonarTS Server is up", e);
+      throw new RuntimeException(e);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define and throw a dedicated exception instead of using a generic one. ([squid:S00112](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00112))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-351&issues=AWD5U5d9EFfjxkQcPdce&open=AWD5U5d9EFfjxkQcPdce)

[ISSUE_KEY]: AWD5U5d9EFfjxkQcPdce